packages/jest-mock/src/index.ts

```diff
@@ -177,17 +177,13 @@ export interface Replaced<T = unknown> {
    * Restore property to its original value known at the time of mocking.
    */
   restore(): void;
-
   /**
    * Change the value of the property.
    */
   replaceValue(value: T): this;
 }
 
-type ReplacedPropertyRestorer<
-  T extends object,
-  K extends PropertyLikeKeys<T>,
-> = {
+type ReplacedPropertyRestorer<T extends object, K extends keyof T> = {
```

Comment:
I think that this should align with:

```ts
function replaceProperty<T, K extends keyof T>(obj: T, key: K, value: T[K]): ReplaceProperty<T[K]>;
```

in the types packages.

Currently (without this change) getting a typescript error:

```
node_modules/@jest/environment/build/index.d.ts:401:26 - error TS2430: Interface 'JestImportMeta' incorrectly extends interface 'ImportMeta'.
  The types of 'jest.replaceProperty' are incompatible between these types.
    Type '<T extends object, K extends Exclude<keyof T, keyof { [K in keyof T as Required<T>[K] extends ClassLike ? K : never]: T[K]; } | keyof { [K in keyof T as Required<T>[K] extends FunctionLike ? K : never]: T[K]; }>, V extends T[K]>(object: T, propertyKey: K, value: V) => Replaced<...>' is not assignable to type '<T, K extends keyof T>(obj: T, key: K, value: T[K]) => ReplaceProperty<T[K]>'.
      Types of parameters 'object' and 'obj' are incompatible.
        Type 'T' is not assignable to type 'object'.

401 export declare interface JestImportMeta extends ImportMeta {
                             ~~~~~~~~~~~~~~

  node_modules/@types/jest/index.d.ts:331:30
    331     function replaceProperty<T, K extends keyof T>(obj: T, key: K, value: T[K]): ReplaceProperty<T[K]>;
                                     ~
    This type parameter might need an `extends object` constraint.
```

So basically one of the issues is fixed here by not using `PropertyLikeKeys` anymore but `T extends object` is still incompatible with `T` so should likely also become just `T` here.

Reply:
Hm.. I think it would be better to have `T extends object` in `@types/jest`. It looks right to have a constrain here.

Reply:
As long as it's aligned I guess. But I don't think `object` is a good choice because `object` means "all non-primitive" types. So any class instances would also match that, for example. I don't know jest well enough to suggest what it should be but perhaps `Record<.., ..>` would be a better fit?

Reply:
If all non-primitive types are objects and if objects can have properties, I think it is alright to assume that these objects can have properties and that these properties might get replaced. Sometimes people add extra properties on arrays or functions. For instance, `expect()` has `expect.not.stringContaining()` which is the property of that function.

Reply:
Just opened https://github.com/DefinitelyTyped/DefinitelyTyped/pull/64936